optimize listpack for stream usage

yihuang opened the pull request:
See issue #6279.

> EDITED for final solution and benchmark numbers.

## Solution

- Pre-allocate `min(stream-node-max-bytes, 4096)` bytes of memory for new listpack.
- When current listpack reached `stream-node-max-entries` limit, reallocate the listpack to exact memory it needs.

## Result
```
$ ./redis-server-new --save no
$ redis-benchmark -P 100 -n 1000000 xadd test '*' a 1 b 2
====== xadd test * a 1 b 2 ======
  1000000 requests completed in 1.48 seconds
  50 parallel clients
  3 bytes payload
  keep alive: 1

0.14% <= 1 milliseconds
0.25% <= 3 milliseconds
0.29% <= 4 milliseconds
0.36% <= 5 milliseconds
0.93% <= 6 milliseconds
59.04% <= 7 milliseconds
93.64% <= 8 milliseconds
98.86% <= 9 milliseconds
99.81% <= 10 milliseconds
100.00% <= 10 milliseconds
675675.69 requests per second
```

```
$ ./redis-server-old --save no
$ redis-benchmark -P 100 -n 1000000 xadd test '*' a 1 b 2
====== xadd test * a 1 b 2 ======
  1000000 requests completed in 1.68 seconds
  50 parallel clients
  3 bytes payload
  keep alive: 1

0.01% <= 1 milliseconds
0.04% <= 3 milliseconds
0.05% <= 4 milliseconds
0.33% <= 5 milliseconds
0.35% <= 7 milliseconds
62.48% <= 8 milliseconds
93.05% <= 9 milliseconds
98.55% <= 10 milliseconds
99.32% <= 11 milliseconds
99.98% <= 12 milliseconds
100.00% <= 12 milliseconds
595592.62 requests per second
```

antirez commented:
Hello @yihuang, that's a great idea conceptually, however this breaks the backward compatibility AFAIK, because we serialized listpacks as-they-are on disk. However there is maybe something similar that we could do? LIke over-allocating the listpack in the first malloc() call, to see if the next calls to realloc will do nothing, so we can avoid all the copying. Not sure if this works well because it depends on the malloc implementation.

yihuang commented:
It only adds 4 bytes at the reverse direction of listpack pointer, other than that, the content of memory is still the same, I don't find the exact code to do the serialization, but as long as it's something like ``memcpy(lp, lpGetTotalBytes(lp))``, there is not an issue AFAIK.

antirez commented:
@yihuang what I mean is, after this patch is applied, will Redis be able to read old RDB files and the contrary? It looks like the patch adds a new field in the listpack structure so it is no longer backward compatible.

yihuang commented:
@antirez I just checked the ``rdbSave`` and ``rdbLoad``, I think the ``rdbSave`` part is unaffected, but ``rdbLoad`` need to copy memory once for the new format, seems like a good tradeoff to make. I've commited this solution, and tested for simple cases.

Another option would be to reserve one bit out of current bytes field, so we can support both formats with or without capability field. But we have to pray for nobody have ever created a listpack more than 2G bytes.

oranagra commented:
@antirez you may recall, back when we worked on the specification of listpack, my feedback was that when implementing, we should take care to avoid doing unnecessary double memory copying.

For instance, when you have a listpack of 1mb, and you want to insert 200 bytes in the middle, normally you'll do a realloc (which is most likely being a NOP since you already have a block that is large enough), and then do a memmove to shift half a megabyte to the right, and make room for the new bytes.
However, in another case, you may fall on a case where realloc is not a NOP, and will actually allocate a new block (realloc will implicitly memcpy the old content to the new allocation), and now we memmove half a megabyte, and end up copying parts of the memory twice.

What can be done, in both **insertions and deletions**, is to look at malloc_usable_size and other techniques to see if re-allocation is at all needed, if not needed, we can either realloc, or not realloc, it doesn't really matters.
However, when realloc is needed, we better not use it, and instead just allocate a new block, do two memory copies (to copy the head and tail parts), and no memmove, and free the old allocation. (we save both memory bandwidth, and also memmove can be slower than memcpy in many scenarios).

This PR (if i understand it correctly) is actually a subset of this idea, and the portion of it that changes the listpack (and RDB) format is actually not needed.

yihuang commented:
@oranagra I just amended the patch to use ``zmalloc_size``, it's extremely simple now. Thanks for suggestion.

zuiderkwast commented on src/listpack.c:
```diff
@@ -656,7 +656,7 @@ unsigned char *lpInsert(unsigned char *lp, unsigned char *ele, uint32_t size, un
 
     /* Realloc before: we need more room. */
-    if (new_listpack_bytes > old_listpack_bytes) {
+    if (new_listpack_bytes > lp_malloc_size(lp)) {
```
```suggestion
    if (new_listpack_bytes > old_listpack_bytes &&
        new_listpack_bytes > lp_malloc_size(lp)) {
```

because `malloc_size()` can be a little bit slow, depending on implementation

yihuang replied:
I think this optimizes for the case of `LP_REPLACE`?

zuiderkwast reviewed (commented):
This looks like a good and simple optimization which has been forgotten(?).

My suggestion is not important. It can be merged as is IMHO.

yihuang commented:
rebased on recent unstable branch, and re-run the benchmarks:
```
$ ./redis-server-old --save no
$ redis-benchmark -P 100 -n 10000000 xadd test '*' a 1 b 2
====== xadd test * a 1 b 2 ======
  10000000 requests completed in 16.58 seconds
  50 parallel clients
  3 bytes payload
  keep alive: 1

0.00% <= 1 milliseconds
0.00% <= 2 milliseconds
0.01% <= 3 milliseconds
0.03% <= 4 milliseconds
0.13% <= 5 milliseconds
0.15% <= 6 milliseconds
0.18% <= 7 milliseconds
60.82% <= 8 milliseconds
98.50% <= 9 milliseconds
99.83% <= 10 milliseconds
99.99% <= 11 milliseconds
100.00% <= 11 milliseconds
603027.19 requests per second
$ ./redis-server-new --save no
$ redis-benchmark -P 100 -n 10000000 xadd test '*' a 1 b 2
====== xadd test * a 1 b 2 ======
  10000000 requests completed in 14.55 seconds
  50 parallel clients
  3 bytes payload
  keep alive: 1

0.00% <= 3 milliseconds
0.04% <= 4 milliseconds
0.11% <= 5 milliseconds
0.13% <= 6 milliseconds
64.91% <= 7 milliseconds
98.82% <= 8 milliseconds
99.86% <= 9 milliseconds
100.00% <= 10 milliseconds
100.00% <= 10 milliseconds
687285.19 requests per second
```

oranagra commented:
I wanna state again that there's still a lot that can be done to avoid unnecessary memmove and memcpy, but that's of course not a reason to merge this. 
I see the current version has two main changes::
1. avoid a realloc by using malloc_size. I'm not sure that has any effect since that realloc would be a NOP. 
2. pre-allocate all stream listpacks to their maximum size. This can indeed reduce a lot of reallocs and the memmove inside them, but could come on the expense of memory usage. 

I would like to check two things:
1. Which of these optimizations is the one that resulted in the increased performance (test by disabling the malloc_size optimization). 
2. What's the extra cost in memory usage in various scenarios.

yihuang commented:
> I would like to check two things:
> 
> 1. Which of these optimizations is the one that resulted in the increased performance (test by disabling the malloc_size optimization).

I tried pre-allocate memory only, the performance don't improve.

> 2. What's the extra cost in memory usage in various scenarios.

Added a `lpShrinkToFit` when rotate the listpack, don't have noticeable effect on the benchmark numbers. Only the current listpack has extra memory now, memory usage difference is not noticeable.

oranagra commented:
I'm not sure your last commit solves the memory problem. There could still be cases where over allocated listpacks will remain. 
But if that's not what's providing the speedup, maybe we should revert all the changes to t_stream.c?

It's a bit odd to me that a malloc_zise check before realloc improves performance. Theoretically realloc could have performed similar check implicitly.. 

in your Benchmark you're using jemalloc, right?

yihuang commented:
> I'm not sure your last commit solves the memory problem. There could still be cases where over allocated listpacks will remain.
> But if that's not what's providing the speedup, maybe we should revert all the changes to t_stream.c?
> 

There's only two cases rotate current listpack, another case is when `stream_node_max_bytes` limit is reached, in that case, the listpack is full already.
Before this, the memory wasted could be significant for small entries, for each listpack the amount of memory wasted is:
```
4096(max_bytes) - 100(max_entries) * 10(entry size)
= 3096
```

> It's a bit odd to me that a malloc_zise check before realloc improves performance. Theoretically realloc could have performed similar check implicitly..
> 
> in your Benchmark you're using jemalloc, right?

yes, i think so, I built with `make` on macos:
```
$ ./src/redis-server --version
Redis server v=255.255.255 sha=8e3c27aa:0 malloc=jemalloc-5.1.0 bits=64 build=2a12a71d10e38098
```

yihuang commented:
> It's a bit odd to me that a malloc_zise check before realloc improves performance. Theoretically realloc could have performed similar check implicitly..

I did a simple experiment on this:
```c
#include "jemalloc.h"
#include <stdio.h>

int main() {
    void* p1 = je_malloc(100);
    void* p2 = je_realloc(p1, 10);
    if (p1 != p2) {
        printf("memory moved\n");
    }
}
```

`memory moved` is printed.

oranagra commented:
i was wondering why would the `if` here make any difference:
```c
void *p = malloc(100);
if (malloc_size(p) != 100)
    p = realloc(p, 100);
```
i.e. assuming that realloc isn't gonna allocate any block and copy any memory.

but now re reading this:

> I tried pre-allocate memory only, the performance don't improve.

i understand what i was missing..
without the check of malloc_size, we would have shrunk it back to the original size, and then gradually grow it back as elements are being added.
and without the pre-allocate the malloc_size would not make any difference.
so we have to have both to get the performance gain (sorry for missing it).

> There's only two cases rotate current listpack, another case is when stream_node_max_bytes limit is reached, in that case, the listpack is full already.

there's a third case which is where someone doesn't add any other entries to the listpack and it remains mostly empty.
i suppose the only way for this to happen and cause significant impact is if the user has many small stream keys, and not many records in each.
i suppose this is acceptable, @guybe7 do you concur?

zuiderkwast commented:
Nice. Maybe the top comment can be more clear to explain what is happening: Allocate max bytes when creating listpack to avoid realloc on every XADD. When listpack reaches max number of entries, shrink the allocation.

yihuang commented:
> Nice. Maybe the top comment can be more clear to explain what is happening: Allocate max bytes when creating listpack to avoid realloc on every XADD. When listpack reaches max number of entries, shrink the allocation.

Done, added some comments.

zuiderkwast commented:
> Done, added some comments.

Oh, good, but I was thinking about the PR description itself. :-)

yihuang commented:
https://github.com/redis/redis/pull/6281/files#diff-efa10b4237199b6d2b155595ff54237dae0760c83e0f8115dbbaa2adaae183b1R735-R739
Should we remove these lines? since it'll shrink memory at once when max entries reached, this seems not needed?

yihuang commented:
> > Done, added some comments.
> 
> Oh, good, but I was thinking about the PR description itself. :-)

Updated PR description, added a new change: don't shrink memory immediately when deletion.

guybe7 commented:
@yihuang thanks for your contribution!

my fear is that is stream_node_max_bytes is very very big, one XADD could crash the server.
i think a better strategy is something in that spirit:
```
if ((lp = lp_realloc(lp,closest_power_of_two(new_listpack_bytes)) == NULL) return NULL;
```

i.e. use realloc, but always allocate more than needed so that we reduce significantly the amount of calls to realloc 
(don't forget to cap it by stream_node_max_bytes)

WDYT?

oranagra commented on src/listpack.c:
```diff
@@ -732,12 +732,6 @@ unsigned char *lpInsert(unsigned char *lp, unsigned char *ele, uint32_t size, un
                 old_listpack_bytes-poff-replaced_len);
     }
 
-    /* Realloc after: we need to free space. */
-    if (new_listpack_bytes < old_listpack_bytes) {
-        if ((lp = lp_realloc(lp,new_listpack_bytes)) == NULL) return NULL;
-        dst = lp + poff;
-    }
-
```
i don't think we want to remove these lines.
keep in mind that listpack is a generic data stracture that can be some day used for many things (we intend to replace all our ziplists).
we need to make sure it use efficient in streams, but we don't want other users to suffer from remaining with huge overheads.
in some way we need to let the caller control when to over allocate, and when to trim or avoid trimming.

zuiderkwast replied:
Off topic: Although listpacks are better than ziplists, we can do even better by using the suffix bytes for payload and size bits, rather than only representing the total size. I have an encoding in mind where a 7bit integer is only 1 byte including suffix (0xxxxxxx), a 12bit integer is 2 bytes (10xxxxxx 10xxxxxx), 16bit integer is 3 bytes (1100xxxx xxxxxxxx 1100xxxx), short strings have only 2 bytes overhead with 4 bits for size in prefix and suffix (1101ssss \<data\> 1101ssss) and so on. I'll post it somewhere soon if you're interested.

yihuang replied:
reverted ;D

oranagra commented:
to emphasize what Guy meant, these two lines will cause a panic:
```
config set stream-node-max-bytes 100G
XADD mystream * field1 value1
```

zuiderkwast commented:
> i think a better strategy is something in that spirit:
> 
> ```
> if ((lp = lp_realloc(lp,closest_power_of_two(new_listpack_bytes)) == NULL) return NULL;
> ```

Resizing in exponental steps is more or less what allocators do, for example jemalloc:

![jemalloc-small-sizes](https://user-images.githubusercontent.com/273886/107969234-ca5d3c80-6faf-11eb-84b7-9b08ed50f4c9.png)

> WDYT?

Another very simple idea to avoids the smallest reallocs is to start with a fairly small fixed size such as 128 bytes and then let realloc do its job from there. WDYT?

oranagra commented:
you're right. assuming HAVE_MALLOC_SIZE wasn't disabled in build time, we'll get some stepping for free.
so we can allocate `MIN(1024,stream_node_max_bytes)` (as a difference against someone setting it to a real high value).
then, the check we already have in place now, will do the rest.
i.e. as soon we'll need more than 1024 bytes, asking for 1025, will actually give us 1280.

but the question is if that's aggressive enough. maybe the benchmark will show it's not providing the goods.
in which case we can go with the approach in sdsMakeRoomFor, which is to double the size if it's below 1mb, and add 1mb if above. 

however, there are some platforms for which we build without HAVE_MALLOC_SIZE and some users that disable in on purpose. so maybe using some algorithm like Guy suggested, or like the one in sds.c is better anyway.

yihuang commented:
- Cap the pre-allocation to 4096 which is also the default config value of `stream_node_max_bytes`.
- Reverted the "not shrink memory when deletion"

I guess it's good compromise for now before we make it more complex? :
- For most stream use cases, entries are small, the pre-allocation will be enough for the whole listpack.
- When 4096 is not enough, we also have the default stepping provided by allocator, if enabled.

More sophisticated allocation strategy can be revisited when try to use listpack in more generic cases?

oranagra commented on src/t_stream.c:
```diff
@@ -509,16 +513,29 @@ int streamAppendItem(stream *s, robj **argv, int64_t numfields, streamID *added_
             int64_t count = lpGetInteger(lpFirst(lp));
-            if (count >= server.stream_node_max_entries) lp = NULL;
+            if (count >= server.stream_node_max_entries) {
+                // Shrink extra pre-allocated memory
```
```suggestion
                /* Shrink extra pre-allocated memory */
```

oranagra commented on src/t_stream.c:
```diff
@@ -509,16 +513,29 @@ int streamAppendItem(stream *s, robj **argv, int64_t numfields, streamID *added_
+         * When listpack reaches max number of entries, shrink the allocation.
+         */
+        size_t prealloc = STREAM_LISTPACK_MAX_PRE_ALLOCATE;
+        if (server.stream_node_max_bytes > 0 && server.stream_node_max_bytes < STREAM_LISTPACK_MAX_PRE_ALLOCATE) {
```
```suggestion
        if (server.stream_node_max_bytes > 0 && server.stream_node_max_bytes < prealloc) {
```

oranagra commented on src/t_stream.c:
```diff
@@ -509,16 +513,29 @@ int streamAppendItem(stream *s, robj **argv, int64_t numfields, streamID *added_
+        /* Create the listpack having the master entry ID and fields.
+         * Allocate max bytes when creating listpack to avoid realloc on every XADD.
+         * When listpack reaches max number of entries, shrink the allocation.
+         */
```
```suggestion
         * Pre-allocate some bytes when creating listpack to avoid realloc on
         * every XADD. Since listpack.c uses malloc_size, it'll grow in steps,
         * and won't realloc on every XADD.
         * When listpack reaches max number of entries, we'll shrink the
         * allocation to fit the data. */
```

oranagra commented on src/t_stream.c:
```diff
@@ -43,6 +43,10 @@
 #define STREAMID_STATIC_VECTOR_LEN 8
 
+/* Max pre-allocation for listpack
+ */
```
```suggestion
/* Max pre-allocation for listpack. This is done to avoid abuse of a user
 * setting stream_node_max_bytes to a huge number. */
```

oranagra commented on src/listpack_malloc.h:
```diff
@@ -42,4 +42,5 @@
 #define lp_malloc zmalloc
 #define lp_realloc zrealloc
 #define lp_free zfree
+#define lp_malloc_size zmalloc_size
```
when HAVE_MALLOC_SIZE is not set, there's a difference between zmalloc_size and zmalloc_usable_size
```suggestion
#define lp_malloc_size zmalloc_usable_size
```

oranagra reviewed (commented):
few minor changes, if they look ok to you i think we can merge this.

guybe7 commented on src/t_stream.c:
```diff
@@ -509,16 +513,31 @@ int streamAppendItem(stream *s, robj **argv, int64_t numfields, streamID *added_
+         * When listpack reaches max number of entries, we'll shrink the
+         * allocation to fit the data. */
+        size_t prealloc = STREAM_LISTPACK_MAX_PRE_ALLOCATE;
+        if (server.stream_node_max_bytes > 0 && server.stream_node_max_bytes < prealloc) {
```
can stream_node_max_bytes be 0? if so, we should fix it..
also, i'd like to remind that we have a MIN macro

yihuang replied:
I think `0` means unlimited.

guybe7 replied:
i see that it is possible for it to be 0 (no idea why) but in this case i think prealloc should be 0... it's a bit weird that:
stream_node_max_bytes == 2 -> prealloc = 2
stream_node_max_bytes == 1 -> prealloc = 1
stream_node_max_bytes == 0 -> prealloc = 4096

guybe7 replied:
```c
if (lp == NULL || lp_bytes >= server.stream_node_max_bytes) {
        master_id = id;
        streamEncodeID(rax_key,&id);
        /* Create the listpack having the master entry ID and fields. */
        lp = lpNew();
        lp = lpAppendInteger(lp,1); /* One item, the one we are adding. */
        lp = lpAppendInteger(lp,0); /* Zero deleted so far. */
```
seems like it'll create a new lp on every streamAppendItem... unless i'm missing something

yihuang replied:
According to this code: https://github.com/redis/redis/blob/unstable/src/t_stream.c#L506
I think `0` is more like `+inf`?

yihuang replied:
> seems like it'll create a new lp on every streamAppendItem... unless i'm missing something

`lp` is set to `NULL` only when a new one needs to be created.

guybe7 replied:
oh ok i think i just didn't look in `unstable`... ok then, +1

zuiderkwast replied:
Seems like it can be zero. The `0` (before LONG_MAX) in the code from config.c below is the minimum:

```C
    createSizeTConfig("stream-node-max-bytes", NULL, MODIFIABLE_CONFIG, 0, LONG_MAX, server.stream_node_max_bytes, 4096, MEMORY_CONFIG, NULL, NULL),
```

The only place this config is used is in t_stream.c:

```C
    /* First of all, check if we can append to the current macro node or
     * if we need to switch to the next one. 'lp' will be set to NULL if
     * the current node is full. */
    if (lp != NULL) {
        if (server.stream_node_max_bytes &&
            lp_bytes >= server.stream_node_max_bytes)
        {
            lp = NULL;
```

... so zero means "no limit" here.

oranagra commented:
merged! thanks everyone.
